feat: Add support for `RegExpExtract`/`RegExpExtractAll`

danielhumanmod opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #2708 .

## Rationale for this change

Spark’s RegExpExtract and RegExpExtractAll expressions were previously not accelerated through Comet.
This PR adds full Rust-native implementations and the corresponding Spark–Comet bridging logic, enabling these two UDFs to run on the native engine while preserving Spark-compatible semantics.

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.
-->

## What changes are included in this PR?

1. Spark-side serde / bridge
	•	Added CometRegExpExtract and CometRegExpExtractAll in strings.scala
2. Rust-native UDF implementation
	•	Introduced SparkRegExpExtract and SparkRegExpExtractAll as ScalarUDFImpl.
3. QueryPlanSerde integration
	•	Wired both expressions into the Comet proto conversion pipeline.


<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## How are these changes tested?

- Added new test cases in CometStringExpressionSuite
- Added Rust-side unit tests for core extract and extract_all semantics.
- All existing string expression tests continue to pass.

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
5. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

danielhumanmod committed:
prototype

danielhumanmod committed:
refactor strings.scala

danielhumanmod committed:
test, format and configs

coderfender commented on native/spark-expr/src/string_funcs/regexp_extract.rs:
```diff
@@ -0,0 +1,401 @@
+    fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType> {
+        // regexp_extract always returns String
+        Ok(DataType::Utf8)
+    }
```
@danielhumanmod , could we verify we dont return a LargeUtf8 as well ?

danielhumanmod committed:
make regexp_extract more align with spark's behavior

martin-g commented on native/spark-expr/src/string_funcs/regexp_extract.rs:
```diff
@@ -0,0 +1,452 @@
+fn extract_group(text: &str, regex: &Regex, idx: usize) -> Result<String> {
+    match regex.captures(text) {
+        Some(caps) => {
+            // Spark behavior: throw error if group index is out of bounds
```
AFAIS Spark returns an empty string when there is no such group - https://github.com/apache/spark/blob/branch-3.5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala#L809-L821

martin-g commented on native/spark-expr/src/string_funcs/regexp_extract.rs:
```diff
@@ -0,0 +1,452 @@
+
+        // Compile regex once
+        let regex = Regex::new(&pattern_str).map_err(|e| {
+            internal_datafusion_err!("Invalid regex pattern '{}': {}", pattern_str, e)
```
```suggestion
            exec_err!("Invalid regex pattern '{}': {}", pattern_str, e)
```
Invalid regex is an user error.

danielhumanmod commented on native/spark-expr/src/string_funcs/regexp_extract.rs:
```diff
@@ -0,0 +1,452 @@
+
+        // idx must be a literal int
+        let idx_val = match idx {
+            ColumnarValue::Scalar(ScalarValue::Int32(Some(i))) => *i as usize,
```
Good call, will fix this, thanks!

danielhumanmod commented on native/spark-expr/src/string_funcs/regexp_extract.rs:
```diff
@@ -0,0 +1,401 @@
+            ColumnarValue::Scalar(ScalarValue::Int32(Some(i))) => *i as usize,
+            _ => {
+                return exec_err!("regexp_extract idx must be an integer literal");
+            }
```
Good point. I checked the Spark's implementation, the index there is cast into i32 only. I was wondering if we should follow their behavior

https://github.com/apache/spark/blob/branch-3.5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala#L811C21-L811C33

danielhumanmod committed:
Solve comments (test not yet fixed)
1. more data type support in scala side
2. unify errors as execution ones
3. reduce code duplication
4. negative index check

danielhumanmod committed:
fix regexp_extract_all test failure

danielhumanmod committed:
refactor udf impl

danielhumanmod committed:
fix rust lint

danielhumanmod committed:
minor updates

danielhumanmod commented:
Hey @coderfender @martin-g , I’ve addressed the comments and refactored the Rust implementation based on the pattern in the similar UDF [regexp_replace](https://github.com/apache/datafusion/blob/fc6d0a4287ab8054d9f56c9ce0a13ffc0ec3c170/datafusion/functions/src/regex/regexpreplace.rs#L88) in DataFusion.

Whenever you have time, I’d appreciate another look. Thanks again for the helpful feedback!

acruise commented:
Is there a way to be confident that an arbitrary user-provided regular expression, which Spark expects to be in Java format, is:
1. valid in the intended native version
2. has the same semantics 😅 

e.g. https://github.com/apache/incubator-gluten/issues/7866

andygrove committed:
Merge remote-tracking branch 'apache/main' into regexp-extract-impl

danielhumanmod committed:
fix escape issue in test

danielhumanmod committed:
Merge remote-tracking branch 'apache/main' into regexp-extract-impl

danielhumanmod committed:
Merge branch 'main' into regexp-extract-impl

danielhumanmod commented on spark/src/test/scala/org/apache/comet/CometStringExpressionSuite.scala:
```diff
@@ -391,4 +391,315 @@ class CometStringExpressionSuite extends CometTestBase {
     }
   }
 
+  test("regexp_extract basic") {
```
Also migrated in the latest commit, thanks!

coderfender commented on spark/src/main/scala/org/apache/comet/serde/strings.scala:
```diff
@@ -286,3 +286,83 @@ trait CommonStringExprs {
+      case Literal(_, DataTypes.IntegerType) =>
+        Compatible()
+      case _ =>
+        Unsupported(Some("Only literal group index is supported"))
```
thank you

coderfender reviewed (commented):
thank you for this great PR @danielhumanmod

danielhumanmod committed:
fix format

danielhumanmod committed:
Merge branch 'main' into regexp-extract-impl
